### ISSUE TITLE
Python (continues issue/discussion #3)

### ISSUE THREAD

**Issue by kevinlekiller** [kevinlekiller/mkvrg#7]

**Comment by WhitePeter** [comment]
Yeah, I did it. Now I just need to rename things. Hopefully nothing breaks. But I am quite pleased with my solution. What do you think?

**Comment by kevinlekiller** [comment]
That's a good solution, much simpler than what I had in mind,  just 2 things I'd change, the help (the numbers are not useful anymore) and moving the dict inside the class (I try to avoid global variables, in this case they are only used in this class). I'm not sure if it works but you could also try assigning `choices=` with `sorted(LOGLEVELS.keys(), key=LOGLEVELS.get, reverse=True)`, saves a bit of memory.

**Comment by WhitePeter** [comment]
> I'm not sure if it works but you could also try assigning choices= with sorted(LOGLEVELS.keys(), key=LOGLEVELS.get, reverse=True)

Yes, that works, to, but I figured, we might need it again. A previous version did reuse it, so I wanted to cache this.

BTW, how can I get rid of [this](https://github.com/kevinlekiller/mkvrg/blob/master/mkvrg.py#L59), we set the loglevel, so why is there a default in this place? simply removing the assignment and only leveing `loglevel` in gets me complaints, that there is one too few arguments.

**Comment by kevinlekiller** [comment]
You'll need to refactor a bit, the issue is handler.setLevel(verbosity). Maybe make handler a property and change the level of it in the method set_verbosity, we only have 1 handler anyways, so it should work.

**Comment by WhitePeter** [comment]
Yes, I just noticed, set_verbosity() or now set_loglevel() is only used once. Does that really need to be a method?
Also, it is a bit too long for my taste, how about set_level()? The log part is clear already from the context.

**Comment by WhitePeter** [comment]
Hmm, and I now still get at least one INFO at -v warning

**Comment by kevinlekiller** [comment]
You could do for example, from the main function : log.logger.setLevel(20) log.handler.setLevel(20), but if you move the handler.setLevel into the Log.set_level, then you kill 2 birds with 1 stone, since you just need to call log.set_level(20) and it will set both.

**Comment by kevinlekiller** [comment]
> Hmm, and I now still get at least one INFO at -v warning

The issue is the Log class is the first instanced, since it's required by everything else, but at that point we've not parsed any arguments, so we set the default loglevel to 20 until we've parsed the arguments.

**Comment by WhitePeter** [comment]
That doesn't sound right. We should make sure the argument get parsed first, then. argparse emits its own messages anyway.

**Comment by kevinlekiller** [comment]
That's what I originally wanted, but I wasn't sure if you were set on using the logger for the argument parsing class so I tried to accommodate, but if we both agree, then we should remove logger from the argument parser to make things a bit simpler.

**Comment by kevinlekiller** [comment]
If we change it, then we can do in the main function: Utils, checkArgs, Log, then set the Log instance on Utils.log

**Comment by WhitePeter** [comment]
So, you're saying everything related to logging needs to go from the argparser, right?

**Comment by kevinlekiller** [comment]
I think it would make the code a bit simpler. We could even instanciate Log in the Utils constructor to make the main function a bit simpler.

**Comment by WhitePeter** [comment]
But wait, how can we then set the loglevel at the argparsing stage?

**Comment by kevinlekiller** [comment]
Oops, was going to edit (deleted).

I think it would make the code a bit more simple.

**Comment by WhitePeter** [comment]
Oh crap, how can I reopen this?

**Comment by WhitePeter** [comment]
Ah, found it ;)

**Comment by WhitePeter** [comment]
Yes, but since we need to set verbosity by a method, Log must be instanciated at the argparsing stage already, or am I missing something? Is this maybe a Catch 22 situation? And if so, how to solve it?

**Comment by kevinlekiller** [comment]
Basically, we make a Utils instance in main (utils = Utils()) then we check the args (checkArgs(utils)), then we set the Utils instance log property: utils.log = Log(utils.loglevel)

utils.loglevel is a property we set in checkArgs (self.utils.loglevel = args.loglevel)

Something like that?

**Comment by WhitePeter** [comment]
Um, OK, have to digest this first I think. This OOP stuff still makes my head hurt. ;)

**Comment by kevinlekiller** [comment]
Think of it like a array kind of:

x = [0 =[0 => "z"], 1 = [0 => "a", 1 => "b"]]

x[0][0] is "z"
x[1][0] is "a"
x[1][1] is "b"

class X:
 def __init__(self):
  self.0 = 0()

class 0:
 0 = "z"

class 1:
 0 = "a"
 1 = "b"
 
x = X()
x.0.0 is "z"
x.0.0 is "a"
x.0.1 is "b"

**Comment by kevinlekiller** [comment]
In a more practical example:

class Options:
 maximum_size = 100
 
 def __init__(self):
  # Here I change the Options.maximum_size from 100 to 50
  self.maximum_size = 50

 def change_max_size(self, size):
  self.maximum_size = size

options = Options() # I create a instance of options
print(options.maximum_size) # This will print 50
options.maximum_size = 100 # I set it to 100
options.change_max_size(200) # Here's another way to change the size.

**Comment by WhitePeter** [comment]
You're making it worse, ow. :-D Maybe I just need a break.

BTW, what does __get_ref_loudness have to do in Checkargs()?

**Comment by kevinlekiller** [comment]
It doesn't but I didn't see a better place to put it, I figured it was something that should only be run once, and Checkargs() is erased from memory after it's ran (since we don't store it in a variable), if we put it in Utils() for example, then it will remain in memory until the program exits.

**Comment by WhitePeter** [comment]
OK, but from a logical POV it does not make sense. It makes it really hard for me to get a grasp, what's going on and why.

**Comment by kevinlekiller** [comment]
We might want to rename CheckArgs, since it also checks binaries, maybe call it Setup or something (you're better at names than me).

**Comment by WhitePeter** [comment]
Oh no, it does too many things I think. Let's make smaller objects. And have a real structure. Isn't that the point of OOP? Have something parse the arguments, then something that checks those for sanity and if the program has all it needs. Or check first, if the binaries are in place. If not, why parse arguments? Fail then and there.

**Comment by kevinlekiller** [comment]
Sure, it could also be a good way for you to lean a bit of OOP.

**Comment by WhitePeter** [comment]
Maybe let's make the -v option a special case that gets parsed separately from argparse. That way we get logging before anything else runs. Then binary check, and then the other arguments.

**Comment by kevinlekiller** [comment]
That's a good idea, check the argv in main maybe (if "-v" in argv or something)?

**Comment by WhitePeter** [comment]
Yes, like that, but not today. Need to get up early tomorrow, so I won't touch any more code now, otherwise I won't be able to quit in time.

**Comment by kevinlekiller** [comment]
Alright, I won't be able to do much in the next few days, I'll still be here to assist in the comments.

**Comment by WhitePeter** [comment]
OK. At least, that way I need to do things and make mistakes myself. ;)

**Comment by kevinlekiller** [comment]
That's the best way to go, try things, fail until you succeed, nothing more rewarding (for me at least).

**Comment by WhitePeter** [comment]
So, here are my thoughts on restructuring.
I want this to be as close as possible to the human POV on files and tracks. Also, the goal is to abstract from files and rather use tracks when it comes to multiprocessing. So the queue gets filled if there are tracks to work on, not files.
The way I understand it, the latter is how it works now: if we have, for lack of a better term, "space" left in the queue add a file to it and have bs1770gain analyze its tracks sequentially. I think the better way of doing this would be to add tracks to the queue and have bs1770gain work on them in parallel, regardless if they are in different files or belong to only one.

So here is a non-exhaustive list of what I think is needed for above to work. Define a class `MkxFile` (x to reflect the different types of matroska files). That class should at least contain the file path and maybe the tracks contained in it, not sure about that yet. But the check if it actually is a matroska should go in there as well, if not destroy the object immediately. Or have another broader class `File`, only a subset of that would be of type `MkxFile`, don't know.
Then there will be a class `Track` or maybe `AudioTrack` which should inherit from `MkxFile` the path to the file. But I think `Track` should be sufficient, since it is already clear that we only deal with the audio side of things. I think of instances of `Track` as children of `MkxFile`, to make clearer what I mean. That IMHO reflects closest how matroska files work, one file with an arbitrary number of "subfiles" (streams/tracks) of different types; we only care about the audio type.
That, I think, gives us the abstraction needed to have the queue only "think" in terms of tracks and not directly know about files.
I am still unsure about how to deal with manipulating the tags, though. Wait till all tracks are analyzed or write them to the file just after analyzing that single track in the same queue instance? Could the concurring queue items step on each others toes there, like, two `mkvpropedit` instances trying to write to the same file at the same time? Obviously one of the two would fail due to locking. Or is that taken care of in `multiprocessing`?

What do you think?

**Comment by kevinlekiller** [comment]
> I am still unsure about how to deal with manipulating the tags, though. Wait till all tracks are analyzed or write them to the file just after analyzing that single track in the same queue instance? Could the concurring queue items step on each others toes there, like, two mkvpropedit instances trying to write to the same file at the same time? Obviously one of the two would fail due to locking. Or is that taken care of in multiprocessing?

Multiprocessing forks the process, (basically it runs a second mkvrg.py), the variables in the original mkvrg.py can't be used directly in those forked processes, so the memory is duplicated and the forked process uses the copy of memory to work from, the only thing multiprocessing can do is optionally return variables from the child process when it is done.

This means we either have to make the child process return the xml file location and create a new list which we then go over to apply the tags after multiprocessing is done, or we try to use multithreading, where all our threads have access to the same memory so we can know when 1 file is being used by mkvpropedit and stall a thread to wait for the other thread to complete.

**Comment by WhitePeter** [comment]
> This means we either have to make the child process return the xml file location and create a new list which we then go over to apply the tags after multiprocessing is done, or we try to use multithreading, where all our threads have access to the same memory so we can know when 1 file is being used by mkvpropedit and stall a thread to wait for the other thread to complete.

OK, I think we have established that multithreading does not offer much performance advantage, so we need to do this with multiprocessing. Or do you see a way to overcome the limitations we saw with multithreading. I think, the reason for those are explained [here](https://docs.python.org/2/glossary.html#term-global-interpreter-lock). That part is referenced in the [introduction of the multiprocessing module's documentation](https://docs.python.org/2/library/multiprocessing.html#introduction). But I am not sure, if I understand everything explained there, so maybe you have a look.

**Comment by kevinlekiller** [comment]
Can mkvpropedit maybe accept multiple arguments for track / file, for example :  `mkvpropedit --tags track:1:/tmp/dsfdf track:2:/tmp/asaa "test.mkv"`

That wouldn't solve our issue, but it would mean less subprocess calls.

**Comment by WhitePeter** [comment]
> Can mkvpropedit maybe accept multiple arguments for track / file, for example :  `mkvpropedit --tags track:1:/tmp/dsfdf track:2:/tmp/asaa "test.mkv"`

Yes, that works, I just checked.

**Comment by WhitePeter** [comment]
Good idea, BTW, somehow I thought that was not possible.

**Comment by kevinlekiller** [comment]
I had my doubts, but that's great if it works, it could make things a bit simpler, we could  have a dictionary of files with their tracks / tmp files then after multiprocessing is done, we run multiprocessing again for mkvpropedit.

For example: `dict = {file1.mkv = {1 = /tmp/dfdf.xml, 2 = /tmp/erdsf.xml}, file2.mkv = {1 = /tmp/dksww.xml, 4 = /tmp/skakd.xml}}`

The numbers would be the track numbers.

**Comment by WhitePeter** [comment]
Hmm, I really think that should be put in some kind of object. My thinking was, that the class `Track` would have a member which represents the tags of each individual track and the corresponding tmp_file. Then walk over all tracks of a file and construct the command options that way.

**Comment by WhitePeter** [comment]
OK, I think I am getting a the hang of your code. But I have to ask, why is [CheckArgs()](https://github.com/kevinlekiller/mkvrg/blob/master/mkvrg.py#L170) a class? It is never really instantiated, rather [called like a function](https://github.com/kevinlekiller/mkvrg/blob/master/mkvrg.py#L40) only once. Is there a reason for that, like some implications on `utils` I don't see?

**Comment by kevinlekiller** [comment]
I prefer keeping things (as much as possible) in classes, even if the instance is destroyed after use, it makes the global code cleaner, it also means you can reuse function names if you have to since they don't collide.

**Comment by WhitePeter** [comment]
BTW, found something of interest:
https://stackoverflow.com/questions/471946/how-to-convert-xml-to-json-in-python
https://stackoverflow.com/questions/191536/converting-xml-to-json-using-python

I think that could be the way to go for keeping the original tags and just adding the new ones, provided converting back to XML works. This way, the RG tags could be put in a json object first, instead of XML. Then merge the json we got from mkvinfo with the RG tags, and only after that do the conversion to XML.

**Comment by kevinlekiller** [comment]
It's a shame we can't send json directly to mkvpropedit, would save the extra step.

I was reading on file magic, checking for matroska is pretty easy:

    import os, binascii
    handle = open("test5.mkv", "rb")
    data = handle.read(4)
    handle.close()
    if (binascii.hexlify(data) != "1a45dfa3"):
        print("file is not matroska!")
    

Basically, check if the first 4 bytes of the file are 0x1a 0x45 0xdf 0xa3, as per: http://www.garykessler.net/library/file_sigs.html

**Comment by WhitePeter** [comment]
Cool. But we need to exclude WEBM, because I have in mkvmerge's man page that it does not support tags. So let's check at least another byte.

**Comment by kevinlekiller** [comment]
The 5th byte on test5.mkv is 0xa3, which does not correspond with 0x93 as per the website I linked above. I suppose it varies based on if it's a video or audio or something?

**Comment by WhitePeter** [comment]
Think so, too. Here is where I got the idea with the docstring after you brought up checking bytes, just forgot the link above:
https://matroska.org/technical/specs/index.html

**Comment by WhitePeter** [comment]
the very first part of the table is the important one

**Comment by WhitePeter** [comment]
> The 5th byte on test5.mkv is 0xa3 ...

Hmm, since I just see it in the spec. Are you sure it was the 5th byte you looked at, because the 4th always is 0xa3. Or is there maybe some offset?

**Comment by kevinlekiller** [comment]
Here's the first 16 bytes of test5.mkv: b'1a45dfa3a34286810142f7810142f281'

**Comment by WhitePeter** [comment]
Oh, and there seems to be some nuance to the spec, because EBML element 0x1a45dfa3 may appear multiple times. Look at the "Mu" column.

**Comment by kevinlekiller** [comment]
test1.mkv: b'1a45dfa3934282886d6174726f736b61'
test2.mkv: b'1a45dfa399bf8467277b3a4282886d61'
test3.mkv: b'1a45dfa399bf8467277b3a4282886d61'
test4.mkv: b'1a45dfa399bf840d0cd2e44282886d61'

**Comment by WhitePeter** [comment]
Can you get another 16 bytes? To see if the matroska string is always in the same place? In my example above in the first 32 bytes.

**Comment by kevinlekiller** [comment]
import os, binascii
     handle = open("test5.mkv", "rb")
     handle.seek(24);
     data = handle.read(8)
     handle.close()
     print (binascii.hexlify(data), data)

b'6d6174726f736b61' b'matroska'

**Comment by kevinlekiller** [comment]
Basically, shifted to byte 24 and read 8 bytes.

**Comment by WhitePeter** [comment]
Nice one :)

**Comment by WhitePeter** [comment]
I'll put that instead of the `mkvinfo` check in mkvrg.py ...

**Comment by kevinlekiller** [comment]
Unfortunately it only works for test5.mkv, the location of the word matroska changes:

	import os, binascii
	for i in range(1, 7):
		handle = open("test" + str(i) + ".mkv", "rb")
		handle.seek(24);
		data = handle.read(8)
		handle.close()
		print (binascii.hexlify(data), data)

b'1853806711642129' b'\x18S\x80g\x11d!)'
b'8102428581021853' b'\x81\x02B\x85\x81\x02\x18S'
b'8102428581021853' b'\x81\x02B\x85\x81\x02\x18S'
b'8101428581011853' b'\x81\x01B\x85\x81\x01\x18S'
b'6d6174726f736b61' b'matroska'
b'1853806711643318' b'\x18S\x80g\x11d3\x18'

**Comment by WhitePeter** [comment]
Hmm, or maybe not. :(

**Comment by kevinlekiller** [comment]
You could  maybe do `if "matroska" in handle.read(32)`

**Comment by kevinlekiller** [comment]
import os, binascii
	for i in range(1, 7):
		handle = open("test" + str(i) + ".mkv", "rb")
		data = handle.read(32)
		handle.close()
		print ("matroska" in str(data), str(data))

True b'\x1aE\xdf\xa3\x93B\x82\x88matroskaB\x87\x81\x02B\x85\x81\x02\x18S\x80g\x11d!)'
True b"\x1aE\xdf\xa3\x99\xbf\x84g'{:B\x82\x88matroskaB\x87\x81\x02B\x85\x81\x02\x18S"
True b"\x1aE\xdf\xa3\x99\xbf\x84g'{:B\x82\x88matroskaB\x87\x81\x02B\x85\x81\x02\x18S"
True b'\x1aE\xdf\xa3\x99\xbf\x84\r\x0c\xd2\xe4B\x82\x88matroskaB\x87\x81\x01B\x85\x81\x01\x18S'
True b'\x1aE\xdf\xa3\xa3B\x86\x81\x01B\xf7\x81\x01B\xf2\x81\x04B\xf3\x81\x08B\x82\x88matroska'
True b'\x1aE\xdf\xa3\x93B\x82\x88matroskaB\x87\x81\x02B\x85\x81\x02\x18S\x80g\x11d3\x18'

**Comment by kevinlekiller** [comment]
Maybe 64 to be safe?

**Comment by WhitePeter** [comment]
Yeah why not. Maybe we can get an accurate number later. It must be in that spec somewhere, but I am too lazy to read it all now. ;)

**Comment by kevinlekiller** [comment]
We know the first 4 bytes should be EBML, so we can seek to byte 5:

	import os
	for i in range(1, 7):
		handle = open("test" + str(i) + ".mkv", "rb")
		handle.seek(5)
		data = handle.read(59)
		handle.close()
		print ("matroska" in str(data))

**Comment by WhitePeter** [comment]
But I think we should maybe check the first 4 bytes, because only the, other test makes sense. It is the same `data` anyway.

**Comment by kevinlekiller** [comment]
Yeah, maybe do both checks.

**Comment by kevinlekiller** [comment]
import os
	for i in range(1, 7):
		handle = open("test" + str(i) + ".mkv", "rb")
		data = handle.read(4)
		print(data == b'\x1aE\xdf\xa3')
		data = handle.read(59)
		handle.close()
		print ("matroska" in str(data))

**Comment by kevinlekiller** [comment]
That seems to work.

**Comment by kevinlekiller** [comment]
`b"matroska" in data` works and doesn't need to convert to string

**Comment by WhitePeter** [comment]
Thx, I am just putting it together.

**Comment by WhitePeter** [comment]
Done. Looks good so far.

**Comment by kevinlekiller** [comment]
I found a PHP library which reads matroska files (and tags), I was trying to figure out how it works, I think I got a decent idea, but it's a pain, it had to go through every byte checking if the bytes match those in the page you listed above, there's no order to how those can be done in a matroska so it has to do conditionals on every byte and change byte offset on the file handle every time accordingly. I was hoping I could just write something simple in python just to read the tags, but that's not possible, going through everything is required, to make sure the byte offset is at the right position.

**Comment by WhitePeter** [comment]
Oh my, looks like it's not worth it. I am still convinced that using the json from mkvinfo and conversion to and from xml can work. We should have everything we need to find the right place to put our data into the json object. Then convert it back to XML and feed that to mkvpropedit. Maybe in the beginning just write all tags back with `--tags all:...`.

**Comment by kevinlekiller** [comment]
Yeah if you want to see the nightmare, here it is : https://github.com/JamesHeinrich/getID3/blob/master/getid3/module.audio-video.matroska.php#L497

**Comment by WhitePeter** [comment]
LOL, it's only ~800 lines of code. No biggy. :P

**Comment by kevinlekiller** [comment]
@005a58d, I think that does not work as expected, I think the first conditional is always true.

**Comment by kevinlekiller** [comment]
This might be a way to do it: https://stackoverflow.com/questions/3389574/check-if-multiple-strings-exist-in-another-string

**Comment by WhitePeter** [comment]
> @005a58d, I think that does not work as expected, I think the first conditional is always true.

Yeah, I thought so myself, reverted it already.

**Comment by WhitePeter** [comment]
Didn't use any() though. I think that is overkill for only 2 strings.

**Comment by kevinlekiller** [comment]
Just had a look at your commits, fine work!

I'm still alive, just a bit busy right now.

**Comment by WhitePeter** [comment]
Thanks, I am glad you like it. I've learned a great deal in the past few days. And I have failed, _a lot_. ;)
And I think I might have been sloppy on the way. Let's see, if I can "harden" the class attributes a bit more, like proper accessors and mutators for data attributes, make the latter as private as python allows, hide as much as possible.
You think it is worth it, or should we keep something like that for later?

**Comment by kevinlekiller** [comment]
I think it's a good idea, if you do it now it might be easier than trying to do it later.

**Comment by WhitePeter** [comment]
Huh, somewhere along the way loglevel setting went missing. I've put in some debug messages which I still cannot see at -v debug. :(

**Comment by WhitePeter** [comment]
FYI, I am currently experimenting with the json module to get the old tags. I think, once we get to the stage where we can just add our tags without destroying prior unrelated ones, we have something to work with.
I think this tagging thing even deserves its own module. It can come in quite handy if one wants to manipulate mkx tags of any kind.

**Comment by kevinlekiller** [comment]
Indeed, there's a lack of tools when it comes to that.

Btw, just some quick testing with mkvinfo:

`mkvinfo -z file`

This shows the size of all the EBML entries:


    + EBML head size 40
    |+ EBML version: 1 size 4
    |+ EBML read version: 1 size 4
    |+ EBML maximum ID length: 4 size 4
    |+ EBML maximum size length: 8 size 4
    |+ Doc type: matroska size 11
    |+ Doc type version: 4 size 4
    |+ Doc type read version: 2 size 4
    + Segment, size 1802186950 size 1802186962
    |+ Seek head (subentries will be skipped) size 86
    |+ EbmlVoid (size: 4004) size 4013
    |+ Segment information size 121
    | + Timecode scale: 1000000 size 7
    | + Muxing application: libebml v1.3.4 + libmatroska v1.4.5 size 38
    | + Writing application: mkvmerge v9.4.0 ('Knurl') 64bit size 34
    | + Duration: 2579.953s (00:42:59.953) size 7
    | + Date: Thu Aug 25 02:52:28 2016 UTC size 11
    | + Segment UID: 0x8e 0xd9 0xfe 0x37 0x01 0xbc 0xa4 0x89 0xa3 0x71 0xcb 0xd1 0x79 0xa4 0x04 0x50 size 19
    |+ Segment tracks size 210
    | + A track size 121
    |  + Track number: 1 (track ID for mkvmerge & mkvextract: 0) size 3
    |  + Track UID: 18308047031429458341 size 11
    |  + Track type: video size 3
    |  + Lacing flag: 0 size 3
    |  + MinCache: 1 size 4
    |  + Codec ID: V_MPEG4/ISO/AVC size 17
    |  + CodecPrivate, length 43 (h.264 profile: High @L4.0) size 46
    |  + Default duration: 41.708ms (23.976 frames/fields per second for a video track) size 8
    |  + Video track size 24
    |   + Pixel width: 1916 size 4
    |   + Pixel height: 1076 size 4
    |   + Display width: 1916 size 7
    |   + Display height: 1076 size 7
    | + A track size 45
    |  + Track number: 2 (track ID for mkvmerge & mkvextract: 1) size 3
    |  + Track UID: 13819279113119135370 size 11
    |  + Track type: audio size 3
    |  + Codec ID: A_AC3 size 7
    |  + Default duration: 32.000ms (31.250 frames/fields per second for a video track) size 8
    |  + Audio track size 11
    |   + Sampling frequency: 48000 size 6
    |   + Channels: 6 size 3
    | + A track size 38
    |  + Track number: 3 (track ID for mkvmerge & mkvextract: 2) size 3
    |  + Track UID: 1016340927463145369 size 11
    |  + Track type: subtitles size 3
    |  + Default flag: 0 size 3
    |  + Lacing flag: 0 size 3
    |  + Codec ID: S_TEXT/UTF8 size 13
    |+ EbmlVoid (size: 1147) size 1150
    |+ Cluster size 54904

By seeking, we can get some data, for example "Segment information":

    import binascii
    handle = open(path, "rb")
    # 40 is EBML head size
    # 8 is the amount of bytes between the EBML head and SeekHead, this was the same for all files I tested.
    # 86 is Seek head
    # 4 is amount of bytes between Seek head and EbmlVoid, same for all files I tested
    # 4013 is EbmlVoid size
    handle.seek(40 + 8 + 86 + 4 + 4013)
    data = handle.read(121)
    hex = str(binascii.hexlify(data), 'ascii')
    hex = ':'.join(hex[i:i+2] for i in range(0, len(hex), 2))
    print(data)
    print(hex)
    handle.close()

Here's the output (string representation and hex under):

    b"\x15I\xa9f\xf4*\xd7\xb1\x83\x0fB@M\x80\xa3libebml v1.3.4 + libmatroska v1.4.5WA\x9fmkvmerge v9.4.0 ('Knurl') 64bitD\x89\x84J\x1dw\xc4Da\x88\x06\xdaH\x11;z8\x00s\xa4\x90\x8e\xd9\xfe7\x01\xbc\xa4\x89\xa3q\xcb\xd1y\xa4\x04P"
   
    15:49:a9:66:f4:2a:d7:b1:83:0f:42:40:4d:80:a3:6c:69:62:65:62:6d:6c:20:76:31:2e:33:2e:34:20:2b:20:6c:69:62:6d:61:74:72:6f:73:6b:61:20:76:31:2e:34:2e:35:57:41:9f:6d:6b:76:6d:65:72:67:65:20:76:39:2e:34:2e:30:20:28:27:4b:6e:75:72:6c:27:29:20:36:34:62:69:74:44:89:84:4a:1d:77:c4:44:61:88:06:da:48:11:3b:7a:38:00:73:a4:90:8e:d9:fe:37:01:bc:a4:89:a3:71:cb:d1:79:a4:04:50


As we can see, the `15:49:a9:66` matches with segment info on: https://matroska.org/technical/specs/index.html

Also near the end we have: `73:a4:90:8e:d9:fe:37:01:bc:a4:89:a3:71:cb:d1:79:a4:04:50`, `7a:a4` is SegmentUID according to https://matroska.org/technical/specs/index.html, `8e:d9:fe:37:01:bc:a4:89:a3:71:cb:d1:79:a4:04:50` matches the `0x8e 0xd9 0xfe 0x37 0x01 0xbc 0xa4 0x89 0xa3 0x71 0xcb 0xd1 0x79 0xa4 0x04 0x50` from the mkvinfo output.

Maybe we can get the tags like this, by seeking to the appropriate location.

**Comment by WhitePeter** [comment]
FYI, I am currently experimenting with the json module to get the old tags. I think, once we get to the stage where we can just add our tags without destroying prior unrelated ones, we have something to work with.
I think this tagging thing even deserves its own module. It can come in quite handy if one wants to manipulate mkx tags of any kind.

**Comment by WhitePeter** [comment]
Oops, why did that out twice...

**Comment by WhitePeter** [comment]
Just FYI, I am still alive. I experimented with above code a little and it looked very promising until I tried to run it on python3. It seems there is fundamental change between python2 and 3 when it comes to processing the json data. I still haven't figured out exactly, where it goes wrong. Strictly speaking StringIO should not be necessary for this, so I went without it, because it was the first to complain. The code runs fine without it in python2 but now json.decode() trips over something else.
Having read only a minimum about the differences and incompatibilities between the major python versions I can only assume that it must have something to do with how python3 "thinks" about file I/O.

I'll keep at it, though, but maybe python2 compatibility was not the best of ideas, after all. :(

On the plus side, I am able to store the original tags in a nice dict construct like `{track_id: {tag_name: tag_value}`. Eventually that will become an attribute of MatroskaFile() or even MatroskaTrack(). If only there wasn't this teeny problem of running on the wrong python version. :P But there must be a way and I am going to find it.

**Comment by kevinlekiller** [comment]
Yeah the inconsistencies between python 2 and 3 are annoying.
